# We will use colaaborative filtering to do it

### We will use mean value imputataion from sklearn

In [161]:
#  Import 
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

In [162]:
# Data Read

data = pd.read_csv("Amazon - Movies and TV Ratings.csv", sep =",")

In [163]:
data.shape

(4848, 207)

In [164]:
data.head()

user_id  Movie1  Movie2  Movie3  Movie4  Movie5  Movie6  Movie7  \
0  A3R5OBKS7OM2IR     5.0     5.0     NaN     NaN     NaN     NaN     NaN   
1   AH3QC2PC1VTGP     NaN     NaN     2.0     NaN     NaN     NaN     NaN   
2  A3LKP6WPMP9UKX     NaN     NaN     NaN     5.0     NaN     NaN     NaN   
3   AVIY68KEPQ5ZD     NaN     NaN     NaN     5.0     NaN     NaN     NaN   
4  A1CV1WROP5KTTW     NaN     NaN     NaN     NaN     5.0     NaN     NaN   

   Movie8  Movie9  ...  Movie197  Movie198  Movie199  Movie200  Movie201  \
0     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   
1     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   
2     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   
3     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   
4     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   

   Movie202  Movie203  Movie204  Movie205  Movie206  
0       NaN       NaN       NaN       NaN       NaN  
1       NaN       NaN       NaN       NaN       NaN  
2       NaN       NaN       NaN       NaN       NaN  
3       NaN       NaN       NaN       NaN       NaN  
4       NaN       NaN       NaN       NaN       NaN  

[5 rows x 207 columns]

In [165]:
data.isnull().sum()   ### Number of null values in each column

user_id        0
Movie1      4847
Movie2      4847
Movie3      4847
Movie4      4846
            ... 
Movie202    4842
Movie203    4847
Movie204    4840
Movie205    4813
Movie206    4835
Length: 207, dtype: int64

In [166]:
# Number of ratings in each column

rating_count_per_movie = data.drop("user_id", axis =1).count(axis=0)    
rating_count_per_movie

Movie1       1
Movie2       1
Movie3       1
Movie4       2
Movie5      29
            ..
Movie202     6
Movie203     1
Movie204     8
Movie205    35
Movie206    13
Length: 206, dtype: int64

### Movie with most rating

In [167]:
rating_count_per_movie.idxmax(), rating_count_per_movie.max(), data.Movie127.count()

('Movie127', 2313, 2313)

#### Movie "Movie127" have most rating that is 2313

### Other way to do it

In [168]:
rating_count_per_movie_df = pd.DataFrame(rating_count_per_movie)

In [169]:
rating_count_per_movie_df.head()

0
Movie1   1
Movie2   1
Movie3   1
Movie4   2
Movie5  29

In [170]:
rating_count_per_movie_df.iloc[:,0].sort_values(axis =0, ascending = False)

Movie127    2313
Movie140     578
Movie16      320
Movie103     272
Movie29      243
            ... 
Movie68        1
Movie69        1
Movie145       1
Movie71        1
Movie1         1
Name: 0, Length: 206, dtype: int64

#### Movie "Movie127" have most rating that is 2313

In [171]:
data_info = data.describe()

### Top 5 most rated Movies

In [172]:
data.mean().sort_values(axis =0, ascending = False).head()

Movie1      5.0
Movie55     5.0
Movie131    5.0
Movie132    5.0
Movie133    5.0
dtype: float64

###  Movies with least audience
#### We will take total count of user ratings per movie
#### And which movie is rated by less people will be output

In [173]:
data.drop("user_id", axis =1).count(axis=0).sort_values(axis =0, ascending = True).head()    

Movie1      1
Movie71     1
Movie145    1
Movie69     1
Movie68     1
dtype: int64

In [174]:
#data = data.fillna(0)
#data.shape

In [175]:
data['user_id'].unique().shape[0]

4848

In [176]:
data.isnull().sum().any()

True

In [177]:
from sklearn.preprocessing import LabelEncoder

In [178]:
from sklearn.preprocessing import LabelEncoder
lenc = LabelEncoder()
lenc.fit(data['user_id'])
data['user_id'] = lenc.transform(data['user_id'])

In [179]:
(data.dtypes == object).any()

False

### Prepare data in the format for recommender system

In [180]:
data.head()

user_id  Movie1  Movie2  Movie3  Movie4  Movie5  Movie6  Movie7  Movie8  \
0     3549     5.0     5.0     NaN     NaN     NaN     NaN     NaN     NaN   
1     4200     NaN     NaN     2.0     NaN     NaN     NaN     NaN     NaN   
2     3339     NaN     NaN     NaN     5.0     NaN     NaN     NaN     NaN   
3     4689     NaN     NaN     NaN     5.0     NaN     NaN     NaN     NaN   
4      458     NaN     NaN     NaN     NaN     5.0     NaN     NaN     NaN   

   Movie9  ...  Movie197  Movie198  Movie199  Movie200  Movie201  Movie202  \
0     NaN  ...       NaN       NaN       NaN       NaN       NaN       NaN   
1     NaN  ...       NaN       NaN       NaN       NaN       NaN       NaN   
2     NaN  ...       NaN       NaN       NaN       NaN       NaN       NaN   
3     NaN  ...       NaN       NaN       NaN       NaN       NaN       NaN   
4     NaN  ...       NaN       NaN       NaN       NaN       NaN       NaN   

   Movie203  Movie204  Movie205  Movie206  
0       NaN       NaN       NaN       NaN  
1       NaN       NaN       NaN       NaN  
2       NaN       NaN       NaN       NaN  
3       NaN       NaN       NaN       NaN  
4       NaN       NaN       NaN       NaN  

[5 rows x 207 columns]

In [181]:
'''
import numpy as np
from sklearn.impute import SimpleImputer
imp = SimpleImputer(missing_values=np.nan, strategy='most_frequent')
new_data = imp.fit_transform(Df.drop("user_id", axis =1))
'''

'\nimport numpy as np\nfrom sklearn.impute import SimpleImputer\nimp = SimpleImputer(missing_values=np.nan, strategy=\'most_frequent\')\nnew_data = imp.fit_transform(Df.drop("user_id", axis =1))\n'

In [182]:
'''
data.user_id.shape[0]
np.array(data.user_id).shape
np.reshape(np.array(data.user_id),(data.user_id.shape[0],1)).shape
new_data_res = np.concatenate([np.reshape(np.array(data.user_id),(data.user_id.shape[0],1)) , new_data], axis =1)
new_data_res = pd.DataFrame(new_data_res)
new_data_res.columns = data.columns
'''

'\ndata.user_id.shape[0]\nnp.array(data.user_id).shape\nnp.reshape(np.array(data.user_id),(data.user_id.shape[0],1)).shape\nnew_data_res = np.concatenate([np.reshape(np.array(data.user_id),(data.user_id.shape[0],1)) , new_data], axis =1)\nnew_data_res = pd.DataFrame(new_data_res)\nnew_data_res.columns = data.columns\n'

##  Code for data prepration

In [183]:
# Mapping every id with every movie and their rating
# So every id will have 206 movie entries and 206 ratings

lst_with_all_possible_combination_of_user_and_movie_ratings = []
all_movie_ids = data.drop(["user_id"], axis =1).columns

for indd,roww in data.iterrows():
    # i is index and j is df of single row
    #print(j.Movie1)
    temp1 = roww.user_id
    
    for k in all_movie_ids:
        #print(k)
        lst_with_all_possible_combination_of_user_and_movie_ratings.append([temp1,k, roww[k]])
    #print(temp2)
    
pd.DataFrame(lst_with_all_possible_combination_of_user_and_movie_ratings)    
    
    

0         1    2
0       3549.0    Movie1  5.0
1       3549.0    Movie2  5.0
2       3549.0    Movie3  NaN
3       3549.0    Movie4  NaN
4       3549.0    Movie5  NaN
...        ...       ...  ...
998683  4242.0  Movie202  NaN
998684  4242.0  Movie203  NaN
998685  4242.0  Movie204  NaN
998686  4242.0  Movie205  NaN
998687  4242.0  Movie206  5.0

[998688 rows x 3 columns]

In [184]:
Df = pd.DataFrame(lst_with_all_possible_combination_of_user_and_movie_ratings)
Df.columns = ["user_id", "movie_id", "ratings"]
Df.head()

user_id movie_id  ratings
0   3549.0   Movie1      5.0
1   3549.0   Movie2      5.0
2   3549.0   Movie3      NaN
3   3549.0   Movie4      NaN
4   3549.0   Movie5      NaN

## Using mean value imputer for data filling

In [185]:
import numpy as np
from sklearn.impute import SimpleImputer
imp = SimpleImputer(missing_values=np.nan, strategy='mean')

In [186]:
# Imputing only ratings column

new_data = imp.fit_transform(Df.drop(["user_id", "movie_id"], axis =1))

In [187]:
Df.user_id.shape, new_data.shape
Df.head()

user_id movie_id  ratings
0   3549.0   Movie1      5.0
1   3549.0   Movie2      5.0
2   3549.0   Movie3      NaN
3   3549.0   Movie4      NaN
4   3549.0   Movie5      NaN

## Concat new movie ratings with user_id and movie_id columns

In [188]:
new_data_res = np.concatenate([np.reshape(np.array(Df.drop("ratings", axis =1)),(Df.user_id.shape[0],2)) , new_data], axis =1)

In [189]:
new_data_res.shape

(998688, 3)

In [190]:
new_data_res = pd.DataFrame(new_data_res)
new_data_res.columns = Df.columns
new_data_res.head()

user_id movie_id ratings
0    3549   Movie1       5
1    3549   Movie2       5
2    3549   Movie3  4.3856
3    3549   Movie4  4.3856
4    3549   Movie5  4.3856

In [191]:
trans_Df = new_data_res.dropna()

In [192]:
trans_Df.shape

(998688, 3)

#### Label encode movie id

In [193]:
le = LabelEncoder()
le.fit(trans_Df['movie_id'])
trans_Df['movie_id'] = le.transform(trans_Df['movie_id'])

In [194]:
from sklearn.model_selection import train_test_split
train, test = train_test_split(trans_Df, test_size = .3)

In [195]:
train.head()

user_id  movie_id ratings
672495    2634        15  4.3856
219195     167        23  4.3856
384496    1481         3  4.3856
125224    2513        93  4.3856
272070    1401        58  4.3856

In [196]:
training_matrix = np.zeros(data.shape)
test_matrix = np.zeros(data.shape)
training_matrix.shape

(4848, 207)

In [197]:
for line in train.itertuples():
    print(line)
    break

Pandas(Index=672495, user_id=2634.0, movie_id=15, ratings=4.3856)


In [198]:
for line in train.itertuples():
    training_matrix[int(line[1]), line[2]] = line[3]
training_matrix  

array([[0.    , 0.    , 0.    , ..., 4.3856, 0.    , 0.    ],
       [4.3856, 0.    , 4.3856, ..., 4.3856, 0.    , 0.    ],
       [4.3856, 0.    , 4.3856, ..., 4.3856, 4.3856, 0.    ],
       ...,
       [4.3856, 4.3856, 0.    , ..., 4.3856, 4.3856, 0.    ],
       [0.    , 4.3856, 4.3856, ..., 4.3856, 4.3856, 0.    ],
       [0.    , 4.3856, 0.    , ..., 4.3856, 0.    , 0.    ]])

In [199]:
for line in test.itertuples():
    test_matrix[int(line[1]), line[2]] = line[3]
test_matrix[test_matrix>4]

array([4.3856, 4.3856, 4.3856, ..., 4.3856, 4.3856, 4.3856])

In [200]:
i, j = np.where(np.isclose(test_matrix, 5))
i, j 

(array([   4,   17,   23, ..., 4834, 4843, 4844], dtype=int64),
 array([ 67,  31,   5, ..., 197,  31,  25], dtype=int64))

In [201]:
test_matrix[6,187]

0.0

In [202]:
from sklearn.metrics.pairwise import pairwise_distances
user_similarity = pairwise_distances(training_matrix, metric='cosine')
item_similarity = pairwise_distances(training_matrix.T, metric='cosine')

In [203]:
def predict(ratings, similarity, type='user'):
    if type == 'user':
        mean_user_rating = ratings.mean(axis=1)
        #print("mean_user_rating", ratings.shape, mean_user_rating[:, np.newaxis].shape)
        #You use np.newaxis so that mean_user_rating has same format as ratings
        ratings_diff = (ratings - mean_user_rating[:, np.newaxis]) 
        #print("ratings_diff", ratings_diff.shape)
        pred = mean_user_rating[:, np.newaxis] + similarity.dot(ratings_diff) / np.array([np.abs(similarity).sum(axis=1)]).T
    elif type == 'item':
        pred = ratings.dot(similarity) / np.array([np.abs(similarity).sum(axis=1)])     
    return pred

In [204]:
item_prediction = predict(test_matrix, item_similarity, type='item')
user_prediction = predict(test_matrix, user_similarity, type='user')

In [205]:
from sklearn.metrics import mean_squared_error
from math import sqrt
def rmse(prediction, ground_truth):
    #print(prediction)
    prediction = prediction[ground_truth.nonzero()].flatten() 
    #print(prediction.shape)
    ground_truth = ground_truth[ground_truth.nonzero()].flatten()
    #print(prediction.shape)
    return sqrt(mean_squared_error(prediction, ground_truth))

In [206]:
#print('Item-based CF RMSE: ' + str(rmse(item_prediction, test_matrix)))
print('User-based CF RMSE: ' + str(rmse(user_prediction, test_matrix)))

User-based CF RMSE: 3.0742551574451964


In [207]:
user_prediction[0,0]

1.5457029451236195

In [208]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler(feature_range= (0,5))
user_prediction_scaled = scaler.fit_transform(user_prediction) 


In [209]:
user_prediction_scaled[user_prediction_scaled>4].shape

(12545,)

In [210]:
user_prediction_scaled[0,0]

3.7414436803394824

In [211]:
#print('Item-based CF RMSE: ' + str(rmse(item_prediction, test_matrix)))
print('User-based CF RMSE: ' + str(rmse(user_prediction_scaled, test_matrix)))

User-based CF RMSE: 2.0174160877217924


## Results iimproved from RMSE (without null value imputation)3.7 to 2(with mean imputation)

In [213]:
## We will use other methods for improving results
## Like a predictor with unsupervised learning

In [214]:
# https://stackoverflow.com/questions/40996051/predict-na-missing-values-with-machine-learning